presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,247 @@
+        }
+    }
+
+    void setReturnType(ReturnType returnType)
+    {
+        this.returnType = returnType;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param ranges the list of minimum and maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    List<ZAddressRange> zOrderCondensedSearchCurve(List<ZValueRange> ranges)
+    {
+        this.ranges = fillInNulls(ranges);
+
+        addressRanges.clear();
+
+        findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength);
+
+        return addressRanges;
+    }
+
+    /**
+     * Searches for and outputs the long addresses within certain ranges in each dimension.
+     *
+     * @param ranges the list of minimum and maximum values of each dimension
+     * @return the list of long addresses within the given value ranges
+     */
+    List<Long> zOrderSearchCurveLongs(List<ZValueRange> ranges)
+    {
+        List<Long> addresses = new ArrayList<>();
+
+        List<ZAddressRange> addressRanges = zOrderCondensedSearchCurve(ranges);
+
+        for (ZAddressRange range : addressRanges) {
+            for (long address = range.getMinAddress(); address <= range.getMaxAddress(); address++) {
+                addresses.add(address);
+            }
+        }
+
+        return addresses;
+    }
+
+    /**
+     * Searches for and outputs the integer addresses within certain ranges in each dimension.
+     *
+     * @param ranges the list of minimum and maximum values of each dimension
+     * @return the list of integer addresses within the given value ranges
+     */
+    List<Integer> zOrderSearchCurveIntegers(List<ZValueRange> ranges)
+    {
+        List<Integer> intAddresses = new ArrayList<>();
+
+        List<ZAddressRange> addressRanges = zOrderCondensedSearchCurve(ranges);
+
+        for (ZAddressRange range : addressRanges) {
+            for (long address = range.getMinAddress(); address <= range.getMaxAddress(); address++) {
+                intAddresses.add((int) address);
+            }
+        }
+
+        return intAddresses;
+    }
+
+    private List<ZValueRange> fillInNulls(List<ZValueRange> ranges)
+    {
+        List<ZValueRange> nonNullValues = new ArrayList<>(ranges.size());
+        for (int index = 0; index < ranges.size(); index++) {
+            ZValueRange range = ranges.get(index);
+
+            Integer minimumValue = range.getMinimumValue();
+            Integer maximumValue = range.getMaximumValue();
+
+            if (minimumValue == null) {
+                minimumValue = 0;
+            }
+            if (maximumValue == null) {
+                maximumValue = 1 << encodingBits.get(index);
```

Comment:
This seems not right.

Suppose the encoding bits is 4 for this dimension. The maximum value for this dimension should be 1111, instead of 10000.